### PR TITLE
Introduce TemplateContextCollector

### PR DESCRIPTION
**Pull request by VincentLanglet**
Extracted from https://github.com/twigstan/twigstan/pull/17

It will be easier if all the Collector have the same signature.

### CODE REVIEW

File: src/PHPStan/Collector/TemplateContextCollector.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * @phpstan-type TemplateData = array{
+ *      startLine: int,
+ *      endLine: int,
+ *      template: string,
+ *      context: string,
+ *  }
+ *
+ * @template-covariant TNodeType of Node
+ * @extends Collector<TNodeType, non-empty-array<TemplateData>>
```
Comment (ruudk) [inline src/PHPStan/Collector/TemplateContextCollector.php:17]:
```suggestion
 * @extends Collector<TNodeType, non-empty-list<TemplateData>>
```
See https://github.com/twigstan/twigstan/pull/21

**Comment by VincentLanglet** [comment]
@ruudk Is it normal that php-cs-fixer fails ?

**Comment by ruudk** [comment]
It happens because it wants to auto commit the fixes for you.

This only works for me, apparently. I'll see how [others](https://github.com/webonyx/graphql-php) have configured this and fix it. For now, could you manually fix it?

**Comment by ruudk** [comment]
Fixed the problem with the code style in main.

**Comment by ruudk** [comment]
I like it 😊 

How can people configure their own collector? Currently, there is no way to control the PHPStan configuration yet. And we need to register those somewhere here: 
https://github.com/twigstan/twigstan/blob/477c4bab2a0e19744d47cd6a5b1b86190565697b/src/Application/PHPStanRunner.php#L45-L60

**Comment by VincentLanglet** [comment]
> I like it 😊
> 
> How can people configure their own collector? 

Currently it's not possible. I would prefer to leave this as a feature request for another PR and first finish https://github.com/twigstan/twigstan/pull/17


> Currently, there is no way to control the PHPStan configuration yet. And we need to register those somewhere here:
> https://github.com/twigstan/twigstan/blob/477c4bab2a0e19744d47cd6a5b1b86190565697b/src/Application/PHPStanRunner.php#L45-L60

I would have in mind something in the twigstan.neon like
```
parameters:
    twigstan:
        directories: [...]

services:
	-
		class: MyCustomCollector
		tags:
			- collector
```
or
```
parameters:
    phpstan:
        services:
	        -
		        class: MyCustomCollector
		        tags:
	    		    - collector
```

And the whole `service` part would be passed like this to phpstan
```
$this->filesystem->dumpFile(
            $tempConfigFile,
            Neon::encode([
                'includes' => [
                    $phpstanConfigurationFile,
                ],
                'parameters' => [
                    'level' => $collectOnly ? null : 8,
                    'customRulesetUsed' => $collectOnly,
                    'twigstan' => [
                        'environmentLoader' => $environmentLoader,
                        'analysisResultJsonFile' => $analysisResultJsonFile,
                    ],
                ],
                'service' => $serviceSection,
            ]),
        );
```

Or we find a way (by changing some key in the existing config) to pass directly the twigstan.neon config (?)
```
$this->filesystem->dumpFile(
            $tempConfigFile,
            Neon::encode([
                'includes' => [
                    $phpstanConfigurationFile,
                    $twigstanConfigurationFile,
                ],
                'parameters' => [
                    'level' => $collectOnly ? null : 8,
                    'customRulesetUsed' => $collectOnly,
                    'twigstan' => [
                        'environmentLoader' => $environmentLoader,
                        'analysisResultJsonFile' => $analysisResultJsonFile,
                    ],
                ],
            ]),
        );
```

**Comment by ruudk** [comment]
> Currently it's not possible. I would prefer to leave this as a feature request for another PR and first finish #17
> 
> >

I agree, let's ship this first and improve later.

Thanks for your contribution 🎉

**Pull request merged**